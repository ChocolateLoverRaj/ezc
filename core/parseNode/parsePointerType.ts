import CoreTokenType from '../CoreTokenType'
import CoreKeyWord from '../CoreKeyWord'
import CoreNodesWithData from './CoreNodesWithData'
import CoreNodeType from './CoreNodeType'
import TryParseNode from './TryParseNode'
import EnumItem from '../EnumItem'
import TryParseNodeResult from './tryParseNodeResult/TryParseNodeResult'

const parsePointerType: TryParseNode<CoreNodesWithData[CoreNodeType.POINTER_TYPE]> =
  async stream => {
    const iterator = stream[Symbol.asyncIterator]()
    const { value, done } = await iterator.next()
    const error: TryParseNodeResult<CoreNodesWithData[CoreNodeType.POINTER_TYPE]> = {
      success: false,
      result: {
        type: { enum: CoreNodeType, id: CoreNodeType.POINTER_TYPE },
        index: 0,
        message: 'Expected ptr keyword',
        subAttempts: undefined
      }
    }
    if (done === true) return error
    if (!(
      value.type.enum === CoreTokenType &&
    value.type.id === CoreTokenType.KEY_WORD &&
    (value.data as EnumItem).enum === CoreKeyWord &&
    (value.data as EnumItem).id === CoreKeyWord.PTR)) {
      return error
    }
    return {
      success: true,
      result: {
        node: {
          type: {
            enum: CoreNodeType,
            id: CoreNodeType.POINTER_TYPE
          },
          data: undefined
        },
        length: 1
      }
    }
  }

export default parsePointerType
