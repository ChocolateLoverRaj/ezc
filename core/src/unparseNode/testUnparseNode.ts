import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from 'util/dist/arrayToAsyncIterable/arrayToAsyncIterable.js'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: ((await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.result as any).node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
