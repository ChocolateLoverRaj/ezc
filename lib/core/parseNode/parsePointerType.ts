import CoreTokenType from '../CoreTokenType'
import KeyWord from '../KeyWord'
import CoreNodesWithData from './CoreNodesWithData'
import CoreNodeType from './CoreNodeType'
import TryParseNode from './TryParseNode'
import EnumItem from '../EnumItem'

const parsePointerType: TryParseNode<CoreNodesWithData[CoreNodeType.POINTER_TYPE]> = async stream => {
  const iterator = stream[Symbol.asyncIterator]()
  const { value, done } = await iterator.next()
  if (done === true) return
  if (!(
    value.type.enum === CoreTokenType &&
    value.type.id === CoreTokenType.KEY_WORD &&
    (value.data as EnumItem).enum === KeyWord &&
    (value.data as EnumItem).id === KeyWord.PTR)) {
    return
  }
  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.POINTER_TYPE
      },
      data: undefined
    },
    length: 1
  }
}

export default parsePointerType
