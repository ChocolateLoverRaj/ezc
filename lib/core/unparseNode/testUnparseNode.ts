import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from '../arrayToAsyncIterable'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: (await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
