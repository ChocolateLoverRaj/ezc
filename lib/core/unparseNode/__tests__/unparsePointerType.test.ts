import parsePointerType from '../../parseNode/parsePointerType'
import testUnparseNode from '../testUnparseNode'

test('ptr', async () => {
  await testUnparseNode(parsePointerType, 'ptr')
})
