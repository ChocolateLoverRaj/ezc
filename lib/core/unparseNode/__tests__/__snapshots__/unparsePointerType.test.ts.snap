// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ptr 1`] = `"ptr"`;
